src/PluginClientBuilder.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Http\Client\Common;
+
+use Http\Client\HttpClient;
+use Http\Client\HttpAsyncClient;
+use Psr\Http\Client\ClientInterface;
+
+/**
+ * Build an instance of a PluginClient with a dynamic list of plugins.
+ *
+ * This builder is immutable, each "build" methods will instead create a new
+ * instance of itself with or without the affected plugin, destroying a
+ * potentially http client built in the new instance.
+ *
+ * @author Baptiste Clavié <clavie.b@gmail.com>
+ */
+final class PluginClientBuilder
+{
+    /** @var Plugin[] */
+    private $plugins = [];
+
+    /** @var ClientInterface | HttpClient | HttpAsyncClient */
+    private $client;
+
+    /** @var ?PluginClient */
+    private $pluginClient;
+
+    /**
+     * @param ClientInterface | HttpClient | HttpAsyncClient $client Client to nest into the plugin client
+     */
+    public function __construct($client)
+    {
+        if (!$client instanceof ClientInterface && !$client instanceof HttpClient && !$client instanceof HttpAsyncClient) {
+            throw new \RuntimeException('You must provide a valid http client');
+        }
+
+        $this->client = $client;
+    }
+
+    public function addPlugin(string $name, Plugin $plugin): self
+    {
+        $builder = new self($this->client);
```

Comment:
agree with mark here too - the builder should not be immutable but stateful. i found no good reference to support that claim, but all builders i have seen act that way. its also perfectly fine to have application code like

```
$builder = new PluginClientBuilder($client);
$builder->addPlugin('foo', new Plugin...);
```